### PR TITLE
Serialization for ColumnSharer and Fixing a bug in serializer in the recursive call

### PR DESCRIPTION
**Pull request by JingJZ160**
### Description
- bug fix for the `_method` argument in the `serializer.py`. 
- Assuming that we are going to keep ColumnSharer in the serialized object, we need to provide the `get_params` and `set_params` method so we can leverage the `ConcreteSerializerMixin`. 

If we serialize ColumnShare into a separate file, this is how it looks like now:
```
{
  "store": {
    "domain": {
      "crim": "NoTransform",
      "zn": "NoTransform",
      "indus": "NoTransform",
      "chas": "NoTransform",
      "nox": "NoTransform",
      "rm": "NoTransform",
      "age": "NoTransform",
      "dis": "NoTransform",
      "rad": "NoTransform",
      "tax": "NoTransform",
      "ptratio": "NoTransform",
      "b": "NoTransform",
      "lstat": "NoTransform",
      "medv": "NoTransform"
    },
    "intent": {
      "crim": "Numeric",
      "zn": "Categoric",
      "indus": "Categoric",
      "chas": "Categoric",
      "nox": "Categoric",
      "rm": "Numeric",
      "age": "Numeric",
      "dis": "Numeric",
      "rad": "Categoric",
      "tax": "Categoric",
      "ptratio": "Categoric",
      "b": "Numeric",
      "lstat": "Numeric",
      "medv": "Numeric",
      "indus_UncommonRemover_Other": "Categoric",
      "indus_2.18": "Categoric",
      "indus_7.87": "Categoric",
      "indus_8.14": "Categoric",
      "indus_6.91": "Categoric",
      "indus_5.13": "Categoric",
      "indus_12.83": "Categoric",
      "indus_8.56": "Categoric",
      "indus_10.01": "Categoric",
      "indus_25.65": "Categoric",
      "indus_21.89": "Categoric",
      "indus_19.58": "Categoric",
      "indus_4.05": "Categoric",
      "indus_2.46": "Categoric",
      "indus_3.44": "Categoric",
      "indus_10.59": "Categoric",
      "indus_6.2": "Categoric",
      "indus_4.93": "Categoric",
      "indus_5.86": "Categoric",
      "indus_3.97": "Categoric",
      "indus_9.9": "Categoric",
      "indus_7.38": "Categoric",
      "indus_5.19": "Categoric",
      "indus_18.1": "Categoric",
      "indus_9.69": "Categoric",
      "tax_296": "Categoric",
      "tax_UncommonRemover_Other": "Categoric",
      "tax_222": "Categoric",
      "tax_311": "Categoric",
      "tax_307": "Categoric",
      "tax_233": "Categoric",
      "tax_284": "Categoric",
      "tax_398": "Categoric",
      "tax_270": "Categoric",
      "tax_276": "Categoric",
      "tax_384": "Categoric",
      "tax_432": "Categoric",
      "tax_188": "Categoric",
      "tax_437": "Categoric",
      "tax_403": "Categoric",
      "tax_193": "Categoric",
      "tax_329": "Categoric",
      "tax_224": "Categoric",
      "tax_277": "Categoric",
      "tax_300": "Categoric",
      "tax_330": "Categoric",
      "tax_264": "Categoric",
      "tax_304": "Categoric",
      "tax_287": "Categoric",
      "tax_666": "Categoric",
      "tax_391": "Categoric",
      "ptratio_UncommonRemover_Other": "Categoric",
      "ptratio_17.8": "Categoric",
      "ptratio_18.7": "Categoric",
      "ptratio_15.2": "Categoric",
      "ptratio_21.0": "Categoric",
      "ptratio_19.2": "Categoric",
      "ptratio_17.9": "Categoric",
      "ptratio_19.7": "Categoric",
      "ptratio_18.6": "Categoric",
      "ptratio_20.9": "Categoric",
      "ptratio_19.1": "Categoric",
      "ptratio_21.2": "Categoric",
      "ptratio_14.7": "Categoric",
      "ptratio_16.6": "Categoric",
      "ptratio_16.4": "Categoric",
      "ptratio_17.4": "Categoric",
      "ptratio_13.0": "Categoric",
      "ptratio_17.6": "Categoric",
      "ptratio_18.4": "Categoric",
      "ptratio_19.6": "Categoric",
      "ptratio_20.2": "Categoric"
    }
  },
  "_class": "ColumnSharer",
  "_method": "dict"
}
```

### CODE REVIEW

File: foreshadow/columnsharer.py [hunk 1]
```diff
@@ -42,6 +44,42 @@ def __init__(self, *args, **kwargs):
...
+        Returns:
+            Params for the column_sharer.
+
+        """
+        # Since the users are not supposed to change the acceptable_keys, I'm
+        # inclined to not show them but we do need to show all of them in the
+        # documentation.
+        return {
+            "store": self.store,
+            # "__acceptable_keys": self.__acceptable_keys
```
Comment (cchoquette) [inline foreshadow/columnsharer.py]: Remove

File: foreshadow/columnsharer.py [hunk 2]
```diff
@@ -42,6 +44,42 @@ def __init__(self, *args, **kwargs):
...
+            **params: params to set
+
+        Returns:
+            The column sharer with stores set.
+
+        """
+        for key in params["store"]:
+            self[key] = params["store"][key]
+
+        # for key in params["__acceptable_keys"]:
```
Comment (cchoquette) [inline foreshadow/columnsharer.py]: remove

File: foreshadow/serializers.py [hunk 3]
```diff
@@ -191,7 +191,10 @@ def serialize(self, method=None, **kwargs):
 
         """
         if method is None:
-            method = self.DEFAULT_OPTION
```
Comment (cchoquette) [inline foreshadow/serializers.py]: I made a similar fix. Glad to know I'm not crazy.

File: foreshadow/serializers.py [hunk 4]
```diff
@@ -291,7 +294,13 @@ def dict_deserialize(cls, data):
         else:
             # Cannot use set_params since steps is a required init arg
             # for Pipelines
-            return cls(**params)
+            if issubclass(cls, PipelineSerializerMixin):
+                return cls(**params)
+            else:
```
Comment (cchoquette) [inline foreshadow/serializers.py]: Why was this needed? Is it for ColumnSharer?

Comment (JingJZ160) [reply foreshadow/serializers.py]: This is an issue found by @adithyabsk. We were concerned that the `\_\_init\_\_` method of `Pipeline` has a required keyword argument called `steps` and the usual way of "initializing the instance first (assuming the \_\_init\_\_ has no required keyword argument) and then call set_params()" may not work. @adithyabsk please correct me if I misunderstood it.

File: foreshadow/columnsharer.py (hunk 2 above)
Comment (JingJZ160) [reply foreshadow/columnsharer.py]: will do.

File: foreshadow/columnsharer.py (hunk 1 above)
Comment (JingJZ160) [reply foreshadow/columnsharer.py]: will do.

Comment (JingJZ160) [reply foreshadow/columnsharer.py]: It has been updated.

File: foreshadow/columnsharer.py (hunk 2 above)
Comment (JingJZ160) [reply foreshadow/columnsharer.py]: updated.

File: foreshadow/serializers.py (hunk 4 above)
Comment (cchoquette) [reply foreshadow/serializers.py]: No that makes complete sense and is true. I would add a comment about the required arg for pipelines.

Comment (JingJZ160) [reply foreshadow/serializers.py]: will do.

Comment (JingJZ160) [reply foreshadow/serializers.py]: Done.

Comment (adithyabsk) [reply foreshadow/serializers.py]: @JingJZ160 is right
